### ISSUE TITLE
DistributedDataParallel on multiple GPU nodes slower than one GPU node. 

### ISSUE THREAD

**Issue by YingleiZhang** [pytorch/pytorch#36298]
Hi, i am using PyTorch DistributedDataParallel to train some models and surpisingly. It become slower when i move from one GPU node to 2 GPU nodes.

Here is the example i am running: https://github.com/huggingface/transformers/blob/v2.3.0/examples/run_lm_finetuning.py

Here is my infrastructure:

Single Node: p3.16xlarge
Training time: 36 mins

Two Nodes: p3.16xlarge
Training time: 1 h 45 mins

In both cases, i am using PyTorch distributed data parallel and GPU utilization is almost always be 100%.

I have enabled NCCL_DEBUG=INFO
I copied the nccl output from single node training and multiple node training in this link below.
https://gist.github.com/YingleiZhang/a8df48eb534ba20ff8f26b5309094b55

It seems on single node mode, NCCL are creating a lot of CHANNELs, but not in the multiple node mode. Could this slow speed due to the networking issues between different GPU nodes?

Any help is appreciated.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @xush6528 @osalpekar

**Comment by osalpekar** [comment]
From the NCCL Debug logs, I see that `nranks` is logged as 8 for the single node case and 16 for the 2-nodes case. By 1 GPU node, do you mean the entire p3.16xlarge? If by the 2 GPU node case you mean 2 p3.16xlarge's, it seems like you would be synchronizing across hosts every batch which would be high overhead - you could try syncing every n batches instead. Also just to confirm, your launch script spawns one process per GPU when starting the training?

It is possible that synchronizing with larger gradients during allreduce combined with networking issues could result in high communication overhead that reflects the slower performance.

**Comment by YingleiZhang** [comment]
Thank you for your reply. This is indeed what i thought. I am trying to get some estimation with NCCL folks on the network bandwidth (https://github.com/NVIDIA/nccl/issues/318). 

This training example is based on roberta-base, which has 125M parameters. The model size is about 500MB. In two nodes cases, i think each gpu core is going to have a copy of the model, and it would be 8*2 * 500MB of data getting synchronized across the network. For each batch, consider forward and backward, we should double this number. So likely 16GB of data will be moving across the network.

Do you think this estimation is accurate?

**Comment by osalpekar** [comment]
Synchronization only occurs after the backward pass, so 8*2*500MB = 8GB should be moved across the network each batch.

**Comment by YingleiZhang** [comment]
The network bandwidth is indeed the bottleneck when synchronizing large models.  NCCL folks provide some numbers in the other thread. Basically, the bandwidth for all reduce operation on intra-gpu communication would be much higher than that of inter-node communication. (NCCl folks mentioned 120 GB/s vs 10 GB/s, https://github.com/NVIDIA/nccl/issues/318).

Thanks for the help. I am closing this issue here.
